autogo/shadow.go

```diff
@@ -0,0 +1,158 @@
+//
+// The effect is similar to:
+//   rsync -zarv --include="*/" --include="*.sh" --exclude="*" "$from" "$to"
+// TODO(fejta): investigate just using rsync?
+//
+// The intended use case of this program is with the autogo_generate in //autogo:def.bzl. This rule will clone your primary workspace into an autogo workspace, and then run gazelle to generate rules for go packages.
+//
+// Usage:
+//   autogo -- <ORIGIN_DIR> <DEST_DIR>
+//
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// main calls shadowClone and exits non-zero on failure
+func main() {
+	if n := len(os.Args); n != 3 {
+		log.Printf("Expected 2 args, not %d", n)
+		log.Fatalf("Usage: %s <ORIGIN> <DEST>", filepath.Base(os.Args[0]))
+	}
+	if err := shadowClone(os.Args[1], os.Args[2]); err != nil {
+		log.Fatalf("Failed to clone %s to %s: %v", os.Args[1], os.Args[2], err)
+	}
+}
+
+// shadowClone walks origin to clone the directory structure and link files at the desination.
+func shadowClone(origin, dest string) error {
+	v := visitor{
+		origin: origin,
+		dest:   dest,
+	}
+	return filepath.Walk(v.origin, v.visit)
+}
+
+// action does something with the current file (mkdir or link)
+type action func(origin string, info os.FileInfo, dest string) error
+
+// visitor stores the origin we're cloning from and destionation we clone to.
+type visitor struct {
+	// origin is the path we read to determine what to clone
+	origin string
+	// dest is where we clone
+	dest string
+}
+
+// visit chooses and then performs the right action for the current path
+func (v visitor) visit(path string, info os.FileInfo, verr error) error {
+	act, dest, err := v.choose(path, info, verr)
+	if err != nil {
+		return err
+	}
+	if act == nil {
+		return nil
+	}
+	return act(path, info, dest)
+}
+
+// choose looks at the current path and returns the appropriate action:
+// - testdata directory => skip it and everything under it
+// - directory => create the directory
+// - go file (.go, .s) => create it at dest
+// - other files => do nothing
+func (v visitor) choose(path string, info os.FileInfo, verr error) (action, string, error) {
+	if verr != nil {
+		return nil, "", fmt.Errorf("failed to walk to %s: %v", path, verr)
+	}
+
+	// If origin is /foo/bar and we receive /foo/bar/something/special
+	// Change this to something/special
+	r, err := filepath.Rel(v.origin, path)
+	// First ensure the path is a child of origin
+	if err != nil {
+		return nil, "", fmt.Errorf("%q is not relative to %q: %v", path, v.origin, err)
+	}
+	if r == ".." || strings.HasPrefix(r, "../") {
+		return nil, "", fmt.Errorf("%q is not a child of %q", path, v.origin)
+	}
+
+	// TODO(fejta): handle symlink loops better
+	// Many repos (dep is one) have testdata folders which are
+	// either symlink loops back to some other part of the repo
+	// or else intentionally do not compile.
+	//
+	// The gazelle tool is not robust to packages containing files that fail to compile.
```

Comment:
besides the symlink loop issue, I thought that gazelle handles the testdata/ directory correctly. if not, we should file bugs against gazelle.

Reply:
Gazelle treats `testdata` as data if there are no build files and no buildable .go files. If you have `testdata` directories with .go files that aren't intended to be built, the usual strategy is to add `// +build ignore` to them, give them some extension other than .go, or add `# gazelle:exclude testdata` to the parent build file (not applicable here though).

Reply:
(There are testdata directories out there that are buildable libraries actually used in tests. I think `golang.org/x/crypto` does that. That's why they're not always treated as data.)